Handle partial-success responses for OTLP trace

jmacd opened the pull request:
Fixes #3102.

This constructs an `error` for passing to `otel.Handle()` from any of the new PartialSuccess messages in OTLP v0.19.

[See the specification of this feature](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#partial-success).

dashpole commented on exporters/otlp/otlpmetric/otlpmetricgrpc/client.go:
```diff
@@ -196,9 +198,12 @@ func (c *client) UploadMetrics(ctx context.Context, protoMetrics *metricpb.Resou
 			ResourceMetrics: []*metricpb.ResourceMetrics{protoMetrics},
 		})
+		if resp != nil && resp.PartialSuccess != nil {
+			otel.Handle(otlpinternal.PartialSuccessToError("data points", resp.PartialSuccess.RejectedDataPoints, resp.PartialSuccess.ErrorMessage))
```
```suggestion
			otel.Handle(otlpinternal.PartialSuccessToError("metric data points", resp.PartialSuccess.RejectedDataPoints, resp.PartialSuccess.ErrorMessage))
```
It wouldn't be obvious to me that a data point is necessarily a metric data point.  This makes it clearer.

dashpole commented on exporters/otlp/otlpmetric/otlpmetrichttp/client.go:
```diff
@@ -170,6 +172,25 @@ func (d *client) UploadMetrics(ctx context.Context, protoMetrics *metricpb.Resou
+			}
+			if respProto.PartialSuccess != nil {
+				otel.Handle(otlpinternal.PartialSuccessToError(
+					"data points",
```
```suggestion
					"metric data points",
```
Same as above.

dashpole commented on exporters/otlp/otlpmetric/otlpmetrichttp/client_test.go:
```diff
@@ -269,3 +270,33 @@ func TestStopWhileExporting(t *testing.T) {
+
+	require.Equal(t, 1, len(*errors))
+	require.Contains(t, (*errors)[0].Error(), "partially successful")
+	require.Contains(t, (*errors)[0].Error(), "2 data points rejected")
```
```suggestion
	require.Contains(t, (*errors)[0].Error(), "2 metric data points rejected")
```

dashpole commented on exporters/otlp/otlpmetric/otlpmetricgrpc/client_test.go:
```diff
@@ -329,3 +330,29 @@ func TestFailedMetricTransform(t *testing.T) {
+
+	require.Equal(t, 1, len(*errors))
+	require.Contains(t, (*errors)[0].Error(), "partially successful")
+	require.Contains(t, (*errors)[0].Error(), "2 data points rejected")
```
```suggestion
	require.Contains(t, (*errors)[0].Error(), "2 metric data points rejected")
```

hanyuancheung commented on exporters/otlp/otlpmetric/otlpmetricgrpc/client.go:
```diff
@@ -196,9 +198,12 @@ func (c *client) UploadMetrics(ctx context.Context, protoMetrics *metricpb.Resou
+		resp, err := c.msc.Export(iCtx, &colmetricpb.ExportMetricsServiceRequest{
 			ResourceMetrics: []*metricpb.ResourceMetrics{protoMetrics},
 		})
+		if resp != nil && resp.PartialSuccess != nil {
+			otel.Handle(otlpinternal.PartialSuccessToError("metric data points", resp.PartialSuccess.RejectedDataPoints, resp.PartialSuccess.ErrorMessage))
```
```suggestion
		if resp != nil {
			if resp.PartialSuccess != nil {
				otel.Handle(otlpinternal.PartialSuccessToError("metric data points", resp.PartialSuccess.RejectedDataPoints, resp.PartialSuccess.ErrorMessage))
			}
```
If resp is nil, then `resp.PartialSuccess` will panic because of nil point.

hanyuancheung commented on exporters/otlp/otlpmetric/otlpmetricgrpc/client_test.go:
```diff
@@ -329,3 +330,29 @@ func TestFailedMetricTransform(t *testing.T) {
+	defer func() {
+		_ = mc.stop()
+	}()
+
+	errors := otlpmetrictest.OTelErrors()
+
+	ctx := context.Background()
+	exp := newGRPCExporter(t, ctx, mc.endpoint)
+	require.NoError(t, exp.Export(ctx, testResource, oneRecord))
+
+	defer func() {
+		_ = exp.Shutdown(ctx)
+	}()
+
```
```suggestion
	errors := otlpmetrictest.OTelErrors()
	ctx := context.Background()
	exp := newGRPCExporter(t, ctx, mc.endpoint)
	require.NoError(t, exp.Export(ctx, testResource, oneRecord))
	defer func() {
		_ = exp.Shutdown(ctx)
	}()
```
These blank lines are meaningless, right? Suggest removing them to keep it more readable.

jmacd replied:
:shrug: I personally find blank lines improve readability.

hanyuancheung commented on exporters/otlp/otlptrace/otlptracegrpc/client.go:
```diff
@@ -196,9 +198,12 @@ func (c *client) UploadTraces(ctx context.Context, protoSpans []*tracepb.Resourc
+		resp, err := c.tsc.Export(iCtx, &coltracepb.ExportTraceServiceRequest{
 			ResourceSpans: protoSpans,
 		})
+		if resp != nil && resp.PartialSuccess != nil {
```
Ditto.

jmacd replied:
See https://go.dev/ref/spec#Logical_operators, "The right operand is evaluated conditionally." This is sometimes called "short-circuit logical evaluation", it's common in all programming languages that I'm familiar with.

dmathieu commented on exporters/otlp/internal/partialsuccess.go:
```diff
@@ -0,0 +1,32 @@
+	if errorMessage == "" {
+		errorMessage = "empty message"
+	}
+	return fmt.Errorf("%w: %s (%d %s rejected)", ErrPartialSuccess, errorMessage, itemsRejected, itemPlural)
```
I know this is indirectly tested with all the other tests. But how about adding explicit tests here?

jmacd commented:
Please consider [54d9dcf](https://github.com/open-telemetry/opentelemetry-go/pull/3106/commits/54d9dcf8f17bcdf99519a84557d52b7b18a3758a) a question about readability. I rewrote the code to have better coverage (to pass the CI checks), but I'm not sure it's a good trade. OTOH, I feel it's not worth adding specialized tests to ensure that errors-in-response-handling (protobuf error, I/O error, etc.) are covered. What do you think?

dashpole commented:
I also prefer the lower-code-coverage version for readability.  I would support merging even if it does not pass those checks. Was it failing the `required` check, or the non-required check?

MrAlias reviewed (commented):
Can we pull the changes to the `otlpmetricgrpc` package out of the PR? They will conflict with https://github.com/open-telemetry/opentelemetry-go/pull/3094

jmacd commented:
I'm not sure anyone is interested in seeing this supported for traces -- from my perspective we may as well wait until after #3094. Thanks @MrAlias.

jmacd commented:
I reverted the otlpmetric portion. I also reverted the questionable style in [54d9dcf](https://github.com/open-telemetry/opentelemetry-go/pull/3106/commits/54d9dcf8f17bcdf99519a84557d52b7b18a3758a). (This is no longer a priority for me, however :grin:)

MadVikingGod commented on exporters/otlp/otlptrace/otlptracehttp/client_test.go:
```diff
@@ -348,3 +350,35 @@ func TestStopWhileExporting(t *testing.T) {
+		assert.NoError(t, exporter.Shutdown(context.Background()))
+	}()
+
+	errors := new([]error)
+	otel.SetErrorHandler(otel.ErrorHandlerFunc(func(err error) {
+		*errors = append(*errors, err)
+	}))
+	err = exporter.ExportSpans(ctx, otlptracetest.SingleReadOnlySpan())
+	assert.NoError(t, err)
+
+	require.Equal(t, 1, len(*errors))
+	require.Contains(t, (*errors)[0].Error(), "partially successful")
+	require.Contains(t, (*errors)[0].Error(), "2 spans rejected")
```
```suggestion
	errors := []error{}
	otel.SetErrorHandler(otel.ErrorHandlerFunc(func(err error) {
		errors = append(errors, err)
	}))
	err = exporter.ExportSpans(ctx, otlptracetest.SingleReadOnlySpan())
	assert.NoError(t, err)

	require.Equal(t, 1, len(errors))
	require.Contains(t, errors[0].Error(), "partially successful")
	require.Contains(t, errors[0].Error(), "2 spans rejected")
```

MadVikingGod commented on exporters/otlp/partialsuccess.go:
```diff
@@ -0,0 +1,68 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otlp // import "go.opentelemetry.io/otel/exporters/otlp"
```
I'm conflicted about where this should live.  On the one hand, I would like users to create an ErrorHandler that does something more intelligent than just log a string.  On the other hand, I would prefer not to expose a new API surface needlessly.

If no one has any other objections to this than I can live with it.

jmacd replied:
I'm a fan of structured logging; I would advocate for a backwards-compatible mechanism to accomplish it. The `PartialSuccess` error can implement some future API that enables it to be expressed in a structured way when the producer logs it to a structure-aware logger.

I didn't know where to put it either. At this level in the current package structure, it falls into the top-level package, which feels OK to me.

MadVikingGod replied:
With no objections to this, we can leave it where it is.

MrAlias replied:
I would rather have these live in an internal package, at lease to start.

MrAlias replied:
I'm not sure what a user will be able to do with an error handler other than log the error given they cannot retry the export. And adding this to the public API seems like adding something only OTel developers will use or care about, not users.

MadVikingGod reviewed (approved):
LGTM.  The only thing that needs addressing is the closing of the http.body.

jmacd commented on exporters/otlp/otlptrace/otlptracehttp/client.go:
```diff
@@ -172,6 +174,25 @@ func (d *client) UploadTraces(ctx context.Context, protoSpans []*tracepb.Resourc
+		// Read the partial success message, if any.
+		var respData bytes.Buffer
+		if _, err := io.Copy(&respData, resp.Body); err != nil {
+			return err
```
@MadVikingGod How do you feel about the way I had it in https://github.com/open-telemetry/opentelemetry-go/pull/3106/commits/54d9dcf8f17bcdf99519a84557d52b7b18a3758a? That was the "less readable" form but it did always reach the `body.Close()`statement.

jmacd commented on exporters/otlp/otlptrace/otlptracehttp/client.go:
```diff
@@ -154,28 +156,48 @@ func (d *client) UploadTraces(ctx context.Context, protoSpans []*tracepb.Resourc
 		}
 
-		var rErr error
+		if resp != nil && resp.Body != nil {
+			defer func() {
+				if err := resp.Body.Close(); err != nil {
+					otel.Handle(err)
+				}
+			}()
+		}
+
```
@MadVikingGod PTAL I tried to make the closing the response body and handling of errors more consistent based on your feedback.

MrAlias commented:
There was an unresolved comment: https://github.com/open-telemetry/opentelemetry-go/pull/3106#discussion_r964063731

> I would rather have these live in an internal package, at lease to start.

I think this PR was merged pre-maturely.

MrAlias commented:
> There was an unresolved comment: [#3106 (comment)](https://github.com/open-telemetry/opentelemetry-go/pull/3106#discussion_r964063731)
> 
> > I would rather have these live in an internal package, at lease to start.
> 
> I think this PR was merged pre-maturely.

I'm opening a to address this.
